common/flatpak-context.c

```diff
@@ -489,6 +497,398 @@ flatpak_context_apply_generic_policy (FlatpakContext *context,
+    }
+
+  g_assert (usb_query->rules->len > 0);
+
+  rule_types = g_hash_table_new (g_direct_hash, g_direct_equal);
+  for (size_t i = 0; i < usb_query->rules->len; i++)
+    {
+      FlatpakUsbRule *usb_rule = g_ptr_array_index (usb_query->rules, i);
+      if (!g_hash_table_add (rule_types, GINT_TO_POINTER (usb_rule->rule_type)))
+        {
+          g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE,
+                       _("Multiple USB query rules of the same type is not supported"));
+          return FALSE;
+        }
+    }
+
+  if (g_hash_table_contains (rule_types, GINT_TO_POINTER (FLATPAK_USB_RULE_TYPE_ALL)) &&
+      g_hash_table_size (rule_types) > 1)
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE,
+                   _("'all' must not contain extra query rules"));
+      return FALSE;
+    }
+
+  if (g_hash_table_contains (rule_types, GINT_TO_POINTER (FLATPAK_USB_RULE_TYPE_DEVICE)) &&
+      !g_hash_table_contains (rule_types, GINT_TO_POINTER (FLATPAK_USB_RULE_TYPE_VENDOR)))
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE,
+                   _("USB queries with 'dev' must also specify vendors"));
+      return FALSE;
+    }
+
+  if (out_usb_query)
+    *out_usb_query = g_steal_pointer (&usb_query);
+
+  return TRUE;
+}
+
+void
+flatpak_usb_query_print (FlatpakUsbQuery *usb_query,
+                         GString         *string)
+{
+  g_assert (usb_query != NULL && usb_query->rules != NULL);
+  g_assert (string != NULL);
+
+  for (size_t i = 0; i < usb_query->rules->len; i++)
+    {
+      FlatpakUsbRule *usb_rule = g_ptr_array_index (usb_query->rules, i);
+
+      if (i > 0)
+        g_string_append_c (string, '+');
+
+      flatpak_usb_rule_print (usb_rule, string);
+    }
+}
+
+static void
+flatpak_context_add_usb_query (FlatpakContext  *context,
+                               FlatpakUsbQuery *usb_query)
+{
+  g_autoptr(FlatpakUsbQuery) copy = NULL;
+  g_autoptr(GString) string = NULL;
+
+  g_assert (context != NULL);
+  g_assert (usb_query != NULL && usb_query->rules != NULL);
+
+  copy = flatpak_usb_query_new ();
+
+  for (size_t i = 0; i < usb_query->rules->len; i++)
+    {
+      FlatpakUsbRule *usb_rule = g_ptr_array_index (usb_query->rules, i);
+      g_ptr_array_add (copy->rules, g_memdup2 (usb_rule, sizeof (FlatpakUsbRule)));
+    }
+
+  string = g_string_new (NULL);
+  flatpak_usb_query_print (usb_query, string);
+
+  g_hash_table_insert (context->allowed_usb_devices,
+                       g_strdup (string->str),
+                       g_steal_pointer (&copy));
+}
+
+static void
+flatpak_context_add_nousb_query (FlatpakContext  *context,
+				 FlatpakUsbQuery *usb_query)
+{
+  g_autoptr(FlatpakUsbQuery) copy = NULL;
+  g_autoptr(GString) string = NULL;
+
+  g_assert (context != NULL);
+  g_assert (usb_query != NULL && usb_query->rules != NULL);
+
+  copy = flatpak_usb_query_new ();
+
+  for (size_t i = 0; i < usb_query->rules->len; i++)
```

Comment:
shouldn't this be implemented as `flatpak_usb_query_clone ()` since it's done in two places?

Reply:
done now